### PR TITLE
gpatch: add gpatch binary to avoid conflict with macOS patch

### PR DESCRIPTION
**Pull request by kit-ty-kate**
- [ ] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [ ] Have you built your formula locally with `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `HOMEBREW_NO_INSTALL_FROM_API=1 brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

This ensures there is a simple way to distinguish between macOS patch (/usr/bin/patch) and homebrew gpatch. Macports installs its version as gpatch and so does BSDs but Homebrew seems to be the only that does not do that.
This PR adds a link rather than rename the binary to keep backward compatibility with the old behaviour.

### CODE REVIEW

File: Formula/g/gpatch.rb
Comment (SMillerDev) [inline Formula/g/gpatch.rb:29]:
```suggestion
    bin.install_symlink bin/"patch" => "gpatch"
```

File: Formula/g/gpatch.rb
Comment (SMillerDev) [inline Formula/g/gpatch.rb:27]:
```suggestion
    system "./configure", *std_configure_args
```

**Comment by jhwoodyatt** [comment]
Does this patch work if `gpatch` is installed keg-only?

**Comment by kit-ty-kate** [comment]
is there any chance this can be merged?

**Comment by kit-ty-kate** [comment]
> Does this patch work if `gpatch` is installed keg-only?

this is a good question. Does any of the homebrew maintainers know if it's possible to change the behaviour of a package in a keg-only context?

**Comment by kit-ty-kate** [comment]
not stale. Waiting for answer

**Comment by kit-ty-kate** [comment]
not stale. Waiting for answer

**Comment by kit-ty-kate** [comment]
not stale. Waiting for answer from the maintainers

**Comment by kit-ty-kate** [comment]
not stale. Waiting for answer from the maintainers

**Comment by iMichka** [comment]
I was looking at this and thought we could maybe do the same as gnu-tar and friends:
https://github.com/Homebrew/homebrew-core/blob/master/Formula/g/gnu-tar.rb

This means using `"--program-prefix=g"` and symlinking to libexec/"patch" in cas we need this too?
Also, we need to make sure the 4 formulae that need gpatch are still building

**Review by MikeMcQuaid — COMMENTED** [review]
Thanks for the PR @kit-ty-kate and I'm really sorry for the delays/fighitng with stalebot here.

@Homebrew/core my thoughts, correct me if you disagree here:
- this formula should have always been keg-only because of installing a `patch` binary
- as @iMichka says: making this behave like `gnu-tar` feels like the nicest way to avoid it being keg-only while continuing to provide an optional `patch` binary

If y'all agree (or at least don't disagree) with the above proposal: I'll let @kit-ty-kate know and we can move forward here.

Thanks all (particularly @kit-ty-kate for the patience and grit here!) ❤️

**Comment by kit-ty-kate** [comment]
After looking at how `gnu-tar` works, i'm personally happy with a `gnu-tar`-like solution.

However we should note that it would very much be a breaking change (for the better) and that it would break users who expect `brew install gpatch` to install it as an immediately available `patch`.

Looking at all the packages that currently use `gpatch` it seems that only `quilt` and `dpkg` would need a slight change (from `patch` to `gpatch`) as the path used is explicit. `ncurses` only depends on `gpatch` on linux, which would require no change if we mirror what was done for `gnu-tar`, and `opam` will continue to work just fine (we lookup for both `patch` and `gpatch` in `PATH` and choose the one that we could actually detect as GNU patch)

I've updated this PR to showcase the change, if you're all happy with this proposal. Thank you all!

**Comment by MikeMcQuaid** [comment]
> I've updated this PR to showcase the change, if you're all happy with this proposal. Thank you all!

@kit-ty-kate Yup, this approach makes sense to me and the code does too 👍🏻 If you can use `brew style --fix` to fix up the style and split this into three commits (one per formula): we should be good to go.

**Comment by kit-ty-kate** [comment]
> @kit-ty-kate Yup, this approach makes sense to me and the code does too 👍🏻 If you can use `brew style --fix` to fix up the style and split this into three commits (one per formula): we should be good to go.

done. Should we also bump all their revisions?

**Comment by MikeMcQuaid** [comment]
@kit-ty-kate only need to bump revisions if both people building from source and using bottles would have a broken setup without reinstalling.

**Review by p-linnane — APPROVED** [review]

**Pull request merged**
